src/Routing.jl

```diff
@@ -0,0 +1,220 @@
+
+    # Initialize the list of upstream basins
+    upstream_basin_list = []
+
+    # Iterate over the direct upstream basins of the current basin
+    for up_basin in graph_dict[basin_id]
+        # Add the current upstream basin
+        push!(upstream_basin_list, up_basin)
+
+        # Recursively collect upstream basins of this upstream basin
+        append!(
+            upstream_basin_list,
+            get_upstream_basins(string(up_basin), graph_dict),
+        )
+    end
+
+    # Remove duplicates to ensure each basin appears only once
+    return unique(upstream_basin_list)
+end
+
+function update_state_from_channel!(
+    river_state::RS,
+    channel_model::CM,
+    env::E,
+    start_date::Date,
+    end_date::Date,
+) where {RS <: RiverState, CM <: AbstractChannelModel, E <: Environment}
+    println("Starting channel update")
+
+    km_to_m = 1e3
+
+    output_dir = env.dyn_env.output_dir
+    timeseries_dir = env.dyn_env.timeseries_dir
+
+    routing_lv_basins_dir = env.static_env.routing_lv_basins_dir
+    attributes_df = CSV.read(
+        joinpath(env.static_env.attributes_dir, "attributes.csv"),
+        DataFrame,
+    )
+
+    C, D = channel_model.wave_velocity, channel_model.diffusivity
+    t_max = channel_model.t_max
+
+    all_basin_ids = read_routing_lv_basins(
+        joinpath(routing_lv_basins_dir, "all_basin_ids.txt"),
+    )
+
+    # Iterate over basins in the given routing level
+    for basin_id in all_basin_ids
+        # Get current streamflow in the basin
+        timeseries_df = CSV.read(
+            joinpath(output_dir, "hillslope_basin_$basin_id.csv"),
+            DataFrame,
+        )
+
+        # Allocate streamflow from upstreams
+        up_streamflow = zeros(length(timeseries_df[:, :streamflow]))
+
+        # Iterate over upstreams
+        upstream_basin_list =
+            get_upstream_basins(string(basin_id), env.graph_dict)
+        for up_basin in upstream_basin_list
+            # Read DataFrame with inputs from upstream
+            up_timeseries_df = CSV.read(
+                joinpath(output_dir, "hillslope_basin_$up_basin.csv"),
+                DataFrame,
+            )
+
+            # Get riverine distance until the outlet of the basin from HydroATLAS
+            dist =
+                attributes_df[
+                    attributes_df.HYBAS_ID .== up_basin,
+                    :DIST_MAIN,
+                ][1] -
+                attributes_df[attributes_df.HYBAS_ID .== basin_id, :DIST_MAIN][1]
+            dist *= km_to_m
+            # Sum the runoff and sub-surface runoff columns
+            up_q = up_timeseries_df[:, :streamflow]
+            up_q = reshape(up_q, :, 1)
+
+            # Allocate streamflow array
+            streamflow = similar(up_q)
+
+            # Generate the h(dist,t) function values
+            x = dist[1]
+            distribution = [
+                x / (2 * t * sqrt(π * D * t)) *
+                exp(-((C * t - x)^2 / (4 * D * t))) for t in 1:t_max
```

Comment:
The hillslope convolution ranges from 0:t_max -1. I think either is OK - the contribution from t = 0 is zero, maybe? but we can be consistent